### DISCUSSION TITLE
Query regarding potential introduction of lower versioned frameworks in dotnet runtime containers

### DISCUSSION THREAD

**Discussion by shantanoo-desai** [dotnet/dotnet-docker#3875] [Q&A] [answered]
## Query Description

- I have a bunch of already compiled `.dll` files and do not have the source files for a project that I wish to containerize.
- The ASP.NET runtime required for this application is `v6.0.0`

### Caveat

There are certain `.dll` binaries where they require _backwards compatible_ v5.0.0 framework to run things successfully.

### Trial

My current `Dockerfile` is the following:

```dockerfile
FROM mcr.microsoft.com/dotnet/aspnet:6.0-focal

RUN mkdir -p /usr/share/App/icon.NExT
COPY ./sku.key /usr/share/App/icon.NExT/

COPY ./appsettings.json /icon/DeployServer-Files/

WORKDIR /icon

COPY ./* /icon/DeployServer-Files/
CMD [ "dotnet", "DeployServer-Files/DeployServer.dll" ]
```

which brings up the server based on some Kestrel configuration I provide in the `appsettings.json` file. Upon triggering some specific workflow on a dedicated HTTP API I keep getting the following logs:

```
                         | The specified framework can be found at:
icon42test        | It was not possible to find any compatible framework version
icon42test        | The framework 'Microsoft.AspNetCore.App', version '5.0.0' (x64) was not found.
icon42test        |   - The following frameworks were found:
icon42test        |       6.0.6 at [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
icon42test        | 
icon42test        | You can resolve the problem by installing the specified framework and/or SDK.
```

I understand the distinction of having separation of different runtimes/frameworks/sdks but would there be some form of documentation as to provide lower versioned framework/sdk into the current runtime?

**Comment by richlander** [comment, marked as answer]
Here's an easy pattern to copy other runtimes into your container image.

https://github.com/octokit/octokit.net/blob/main/.devcontainer/Dockerfile#L8-L9

That example is copying the SDK. You don't want that. You'd want:

```Dockerfile
COPY --from=mcr.microsoft.com/dotnet/aspnet:5.0-focal /usr/share/dotnet/shared /usr/share/dotnet/shared
```

I switch to pulling Ubuntu 20.04. This way, you only pull the OS once.

.NET 5 is now out of support, however.

**Reply by shantanoo-desai** [reply]
Thank you. I just found the solution exactly as you mentioned a few minutes ago. I was able to pull two different SDKs into the runtime container and cross-checked it using `dotnet --list-sdks` and `dotnet --list-runtimes`. For posterity my [Answer on StackOverflow query](https://stackoverflow.com/a/72745810/4851126)

**Reply by richlander** [reply]
Excellent. Glad that your problem is now solved.

> I switch to pulling Ubuntu 20.04. This way, you only pull the OS once.

I just realized that while this was good thinking on my part, it doesn't actually apply here. We've stopped updating the .NET 5 images, so the OS layer won't actually match. You will be pulling Ubuntu twice. There isn't a way to avoid that using this technique.
